zig cc: support reading from stdin via "-x LANG -"

motiejus opened the pull request:
echo 'some C program' | $CC -x c -

Is a common pattern to test for compiler or linker features. This PR adds support for reading from non-regular files.

This will make at least one more Go test to pass.

motiejus commented:
The build fails like this:

```
: && /home/ci/deps/zig+llvm+lld+clang-x86_64-linux-musl-0.11.0-dev.971+19056cb68/bin/zig  cc -target x86_64-linux-musl -mcpu=baseline -g -rdynamic CMakeFiles/zig1.dir/zig1.c.o CMakeFiles/zig1.dir/stage1/wasi.c.o -o zig1  -lm && :
ld.lld: error: undefined symbol: wasi_snapshot_preview1_fd_seek
>>> referenced by zig1.c:31166
```

However, I can see this in `wasm.zig`:

```
pub extern "wasi_snapshot_preview1" fn fd_seek(fd: fd_t, offset: filedelta_t, whence: whence_t, newoffset: *filesize_t) errno_t;
```

Looking ...

matu3ba commented on src/main.zig:
```diff
@@ -650,6 +650,44 @@ fn optionalStringEnvVar(arena: Allocator, name: []const u8) !?[]const u8 {
+            break :blk try std.fmt.allocPrint(arena, "tmp" ++ s ++ "{x}", .{rand_int});
+        }
+    };
+    try dst_dir.handle.makePath("tmp");
```
Why is using `zig-cache/tmp/compiler` (next to the build.zig) not an option?

motiejus replied:
It trying to use the local cache directory, if that's what you mean. Can you elaborate/

matu3ba replied:
Ah, I should have phrased this better. The code does not explain, why `tmp` in the `local cache directory` should not be used. Can you add a brief comment?

motiejus replied:
Sorry, I still don't follow your comment. Can you re-phrase?

Perhaps I should move `Compilation.tmpFilePath` out of `Compilation` and re-use the function from everywhere with a `Dir` instead of `Compilation`.

matu3ba replied:
Apologies for my confusing phrasing.
As I understand you, I think yes. `fmt` also is wasting binary space (as of now), so this sounds like the right tradeoff.

If it is next to build.zig, then it sounds logical to reuse the function in Compilation.zig as the only difference is the used suffix (you could use an empty one at the call site).

**However**, if this api is intended to be called from anywhere (not only next to a build.zig), then a separate function sounds okay to me.

motiejus replied:
Your comment just uncovered a bug in my implementation: I asked for an exclusive file, but there was no retry mechanism on collisions. However, it does not have to be exclusive.

I also understood your suggestion! Refactoring and will move to a single function.

matu3ba commented on test/cli.zig:
```diff
@@ -96,18 +100,65 @@ fn exec(cwd: []const u8, expect_0: bool, argv: []const []const u8) !ChildProcess
+fn testZigCCPosix(zig_exe: []const u8, dir_path: []const u8) !void {
+    var tmp = testing.tmpDir(.{});
+    defer tmp.cleanup();
+    try tmp.dir.writeFile("truth.c", "int main() { return 42; }");
```
Why is using child_proess.zig no option here?

motiejus replied:
That's written in https://github.com/ziglang/zig/pull/14462/files#diff-762da4372d6d5407ded5b08627e8ee3400f9687590ab42ab43eb1246f8bf5646R116-R118 : in case that was not clear,  I wasn't able to spawn a `ChildProcess` whose stdin would be custom fd.

Does that explain it?

matu3ba replied:
I'm not sure, if lib/std/child_process should encode specifically all the things one can do with file descriptors to standard streams or where specifically we set the limit, because handling Windows without child_process.zig is extremely verbose.

My PR https://github.com/ziglang/zig/pull/14152 simplifies serializing and deserialization the handle=fd, but is still more verbose (requires child process to have parsing stdin/environment). Would that be an option?
Otherwise it would be nice, if you can leave a review on what I can improve.

andrewrk commented:
Have you considered the alternate implementation using `clang_passthrough_mode` instead of buffering the input to a temporary file?

I don't want to support the use case of stdin as a root source file in any case other than C/C++ compiler compatibility mode.

motiejus commented:
>

motiejus commented:
> Have you considered the alternate implementation using `clang_passthrough_mode` instead of buffering the input to a temporary file?

I ruled it out for a reason I can no longer remember.

This is how I understand it: if the input file is `-`, set `clang_passthrough_mode = true` for the `zig clang` command (which will cause it to read from stdin), and then do some dancing with the filenames from the subsequent steps like linking.

I will look at this again.

> I don't want to support the use case of stdin as a root source file in any case other than C/C++ compiler compatibility mode.

I agree.

motiejus commented:
I started poking at this and remember better now. I did indeed consider the clang passthrough mode; we had this discussion before (around Oct 2022) and you advised me to slurp the file and pass it on to the remaining compilation chain as a regular file. :)

Here was my argument: everything in `src/main.zig`, `src/Compilation.zig` and `src/Cache.zig` assume that all the input files are files. So we have 2 choices:

- add special handling for stdin in those 3 files, or
- slurp stdin and keep assuming everything is a file. Like in this PR.

For a taste on how many things will need to change to adjust this, here is a start:

```diff
diff --git a/src/Compilation.zig b/src/Compilation.zig
index 7d42d3b61..8b351270e 100644
--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -210,7 +210,10 @@ pub const LangToExt = std.ComptimeStringMap(FileExt, .{
 
 /// For passing to a C compiler.
 pub const CSourceFile = struct {
-    src_path: []const u8,
+    src: union(enum) {
+        stdin,
+        path: []const u8,
+    },
     extra_flags: []const []const u8 = &.{},
     /// Same as extra_flags except they are not added to the Cache hash.
     cache_exempt_flags: []const []const u8 = &.{},
```

There are many places in code that will change as a result of this. I can do it, but:
- it will become more complex due to this single case.
- we will lose caching for all compilations that arrive over stdin. This may or may not be a good thing.

Let me know.

motiejus commented:
I can reduce the surface area of this feature this way: allow only `-` as a non-file entry; all other non-files except `-` would be not allowed.

motiejus commented:
Rebased, rewrote the test for the "build system v2".

Also, simplified the change to only account for `-` as stdin. No other streams will be accepted.

andrewrk commented on test/tests.zig:
```diff
@@ -1095,3 +1136,50 @@ pub fn addCases(
+    const max_output_size = 100 * 1024;
+    const result = std.ChildProcess.exec(.{
+        .allocator = allocator,
+        .argv = argv,
+        .cwd = cwd,
+        .max_output_bytes = max_output_size,
+    }) catch |err| {
+        std.debug.print("The following command failed:\n", .{});
+        printCmd(cwd, argv);
+        return err;
+    };
+    switch (result.term) {
+        .Exited => |code| {
+            if (code != expect_code) {
+                std.debug.print(
+                    "The following command exited with error code {}, expected {}:\n",
+                    .{ code, expect_code },
+                );
+                printCmd(cwd, argv);
+                std.debug.print("stderr:\n{s}\n", .{result.stderr});
+                return error.CommandFailed;
+            }
+        },
+        else => {
+            std.debug.print("The following command terminated unexpectedly:\n", .{});
+            printCmd(cwd, argv);
+            std.debug.print("stderr:\n{s}\n", .{result.stderr});
+            return error.CommandFailed;
+        },
+    }
+    return result;
+}
+
+fn printCmd(cwd: []const u8, argv: []const []const u8) void {
+    std.debug.print("cd {s} && ", .{cwd});
+    for (argv) |arg| {
+        std.debug.print("{s} ", .{arg});
+    }
+    std.debug.print("\n", .{});
+}
```
this should not be added

andrewrk commented on test/tests.zig:
```diff
@@ -772,6 +772,47 @@ pub fn addCliTests(b: *std.Build) *Step {
+    // Test author was not able to figure out how to start a child process and
+    // give an fd to it's stdin. fork/exec works, but we are limiting ourselves
+    // to POSIX.
+    if (builtin.os.tag != .windows) {
+        const tmp_path = b.makeTempPath();
+        var dir = std.fs.cwd().openDir(tmp_path, .{}) catch @panic("unhandled");
+        dir.writeFile("truth.c", "int main() { return 42; }") catch @panic("unhandled");
+        var infile = dir.openFile("truth.c", .{}) catch @panic("unhandled");
+
+        const outfile = std.fs.path.joinZ(
+            b.allocator,
+            &[_][]const u8{ tmp_path, "truth" },
+        ) catch @panic("unhandled");
+        const pid_result = std.os.fork() catch @panic("unhandled");
+        if (pid_result == 0) { // child
+            std.os.dup2(infile.handle, std.os.STDIN_FILENO) catch @panic("unhandled");
+            const argv = &[_:null]?[*:0]const u8{
+                b.zig_exe, "cc",
+                "-o",      outfile,
+                "-x",      "c",
+                "-",
+            };
+            const envp = &[_:null]?[*:0]const u8{
+                std.fmt.allocPrintZ(b.allocator, "ZIG_GLOBAL_CACHE_DIR={s}", .{tmp_path}) catch @panic("unhandled"),
+            };
+            const err = std.os.execveZ(b.zig_exe, argv, envp);
+            std.debug.print("execve error: {any}\n", .{err});
+            std.os.exit(1);
+        }
+
+        const res = std.os.waitpid(pid_result, 0);
+        assert(0 == res.status);
+
+        // run the compiled executable and check if it's telling the truth.
+        _ = exec(b.allocator, tmp_path, 42, &[_][]const u8{outfile}) catch @panic("unhandled");
+
+        const cleanup = b.addRemoveDirTree(tmp_path);
+        step.dependOn(&cleanup.step);
+    }
```
This test as is does the following things directly in the configure phase rather than using build steps like it must do in order to play nicely with the other build steps:
 * making a temp dir and writing a file
 * calling fork, execve, and wait
 * calling exec

If you look at the Test Godbolt API code block above, it shows how to use the build system API for testing. It will be a combination of WriteFileStep, and RunStep.

I'm happy to provide assistance in writing this test.

andrewrk reviewed (changes requested):
Thanks for adding this. I have some requested changes. Let me know if you want any help with them.

motiejus commented:
Note to myself: handle this case: https://github.com/ziglang/zig/issues/15338#issuecomment-1516176556

motiejus commented:
I have addressed all the review comments for the code, but not for the tests. Need some help there.

andrewrk commented:
Alright, I have made some executive decisions and finished this patch. The final diff ended up being quite small and tidy.

I removed the dead code added to tests.zig; adding test coverage for this CLI usage can be a separate enhancement. Here is a clue should you wish to pursue this: you'll want to have a Run step with the `stdin` field populated with your C source code, and the `stdio` field set to `check` with some checks against stderr or stdout. If you want to run the created binary, you can use `addOutputArg` on the Run step, and use that with a second Run step with `addFileSourceArg`.

I have tested this locally and force-pushed to your branch, with auto-merge enabled.

motiejus commented on src/main.zig:
```diff
@@ -3020,6 +3020,32 @@ fn buildOutputType(
+            std.crypto.random.int(u64), ext.canonicalName(target_info.target),
+        });
+        try local_cache_directory.handle.makePath("tmp");
+        // Note that in one of the happy paths, execve() is used to switch
+        // to clang in which case any cleanup logic that exists for this
+        // temporary file will not run and this temp file will be leaked.
+        // Oh well. It's a minor punishment for using `-x c` which nobody
+        // should be doing. Therefore, we make no effort to clean up. Using
+        // `-` for stdin as a source file always leaks a temp file.
```
we could minimize the damage by:
- hashing the file contents (while writing to the file, so no additional allocations)
- renaming the filename to the hash.

That way, we avoid a small chance of hash collision (I know) and multiple invocations of `echo | zig cc <...>` produce only one garbage-file.

Thoughts?

andrewrk replied:
OK, sounds reasonable

motiejus replied:
https://github.com/ziglang/zig/pull/15770 and https://github.com/ziglang/zig/pull/15772
